pandas/core/groupby/ops.py

```diff
@@ -735,73 +724,135 @@ def has_dropped_na(self) -> bool:
+    def result_index(self) -> Index:
+        return self.result_index_and_ids[0]
+
+    @property
+    def ids(self) -> np.ndarray:
+        return self.result_index_and_ids[1]
 
     @cache_readonly
-    def result_index(self) -> Index:
+    def result_index_and_ids(self) -> tuple[Index, np.ndarray]:
+        names = self.names
+        codes = [ping.codes for ping in self.groupings]
+        levels = [Index._with_infer(ping.uniques) for ping in self.groupings]
+        obs = [
+            ping._observed or not ping._passed_categorical for ping in self.groupings
+        ]
+        # When passed a categorical grouping, keep all categories
+        for k, (ping, level) in enumerate(zip(self.groupings, levels)):
+            if ping._passed_categorical:
+                levels[k] = level.set_categories(ping._orig_cats)
+
         if len(self.groupings) == 1:
-            return self.groupings[0]._result_index.rename(self.names[0])
+            result_index = levels[0]
+            result_index.name = names[0]
+            ids = ensure_platform_int(codes[0])
+            return result_index, ids
+
+        if any(obs):
+            ob_codes = [code for code, ob in zip(codes, obs) if ob]
+            ob_levels = [level for level, ob in zip(levels, obs) if ob]
+            ob_names = [name for name, ob in zip(names, obs) if ob]
+
+            shape = tuple(len(level) for level in ob_levels)
+            group_index = get_group_index(ob_codes, shape, sort=True, xnull=True)
+            ob_ids, obs_group_ids = compress_group_index(group_index, sort=self._sort)
+            ob_ids = ensure_platform_int(ob_ids)
+            ob_index_codes = decons_obs_group_ids(
+                ob_ids, obs_group_ids, shape, ob_codes, xnull=True
+            )
+            ob_index = MultiIndex(
+                levels=ob_levels,
+                codes=ob_index_codes,
+                names=ob_names,
+                verify_integrity=False,
+            )
 
-        codes = self.reconstructed_codes
-        levels = [ping._result_index for ping in self.groupings]
-        return MultiIndex(
-            levels=levels, codes=codes, verify_integrity=False, names=self.names
-        )
+        if not all(obs):
+            unob_codes = [e for e, o in zip(codes, obs) if not o]
+            unob_levels = [e for e, o in zip(levels, obs) if not o]
+            unob_names = [e for e, o in zip(names, obs) if not o]
+
+            shape = tuple(len(level) for level in unob_levels)
+            unob_ids = get_group_index(unob_codes, shape, sort=True, xnull=True)
+            unob_index = MultiIndex.from_product(unob_levels, names=unob_names)
+
+        if all(obs):
+            result_index = ob_index
+            ids = ensure_platform_int(ob_ids)
+        elif not any(obs):
+            result_index = unob_index
+            ids = ensure_platform_int(unob_ids)
+        else:
+            # Combine unobserved and observed parts of result_index
```

Comment:
Could we do something like

```python
if all(obs):
    result_index, _ = get_ob_index(...)
elif not any(obs):
    result_index, _ = get_unob_index(...)
else:
   ob_index, ob_ids = get_ob_index(...)
   unob_index, unob_ids = get_unob_index(...)
   ...
```

In place of the 2 prior if statements above? IMO that's a little easier to follow in place of 4 `if [not] any/all(objs)`

Reply:
Yea - great idea.